### PR TITLE
Dev/pointer offset

### LINKED ISSUE #89: NREs in pointer inspectors
# XRTK - Mixed Reality Toolkit Bug Report

## Describe the bug

Errors when Right_DefaultControllerPointer(Clone) or Right_ParabolicPointer(Clone) are selected in scene during play mode. UWP or OpenVR mode

## To Reproduce

1. Unity Hub > New project > 3D
2. Modify manifest to include `"com.xrtk.core": "https://github.com/XRTK/XRTK-Core.git#0.1.18"`,
3. Press Play
4. Under MixedRealityPlayspace GO select one of the pointer objects and view their configurations in the inspector.

## Expected behavior

Does not show error. I believe these were rendering a line in previous version of XRTK. I'm not sure what the purpose of these objects are 

## Actual behavior

If I click that Right_DefaultControllerPointer(Clone)
```
NullReferenceException: Object reference not set to an instance of an object
UnityEditor.PropertyHandler.OnGUILayout (UnityEditor.SerializedProperty property, UnityEngine.GUIContent label, System.Boolean includeChildren, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/ScriptAttributeGUI/PropertyHandler.cs:203)
UnityEditor.EditorGUILayout.PropertyField (UnityEditor.SerializedProperty property, UnityEngine.GUIContent label, System.Boolean includeChildren, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/EditorGUI.cs:9341)
UnityEditor.EditorGUILayout.PropertyField (UnityEditor.SerializedProperty property, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/EditorGUI.cs:9325)
XRTK.SDK.Inspectors.UX.Pointers.BaseControllerPointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/BaseControllerPointerInspector.cs:65)
XRTK.SDK.Inspectors.UX.Pointers.LinePointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/LinePointerInspector.cs:40)
UnityEditor.UIElements.InspectorElement+<CreateIMGUIInspectorFromEditor>c__AnonStorey1.<>m__0 () (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorElement.cs:462)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

If I click Right_ParabolicPointer(Clone)
```
NullReferenceException: Object reference not set to an instance of an object
UnityEditor.PropertyHandler.OnGUILayout (UnityEditor.SerializedProperty property, UnityEngine.GUIContent label, System.Boolean includeChildren, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/ScriptAttributeGUI/PropertyHandler.cs:203)
UnityEditor.EditorGUILayout.PropertyField (UnityEditor.SerializedProperty property, UnityEngine.GUIContent label, System.Boolean includeChildren, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/EditorGUI.cs:9341)
UnityEditor.EditorGUILayout.PropertyField (UnityEditor.SerializedProperty property, UnityEngine.GUILayoutOption[] options) (at C:/buildslave/unity/build/Editor/Mono/EditorGUI.cs:9325)
XRTK.SDK.Inspectors.UX.Pointers.BaseControllerPointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/BaseControllerPointerInspector.cs:65)
XRTK.SDK.Inspectors.UX.Pointers.LinePointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/LinePointerInspector.cs:40)
XRTK.SDK.Inspectors.UX.Pointers.TeleportPointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/TeleportPointerInspector.cs:48)
XRTK.SDK.Inspectors.UX.Pointers.ParabolicTeleportPointerInspector.OnInspectorGUI () (at Library/PackageCache/com.xrtk.sdk@2a5ea7c69c675a92e54c82216e78960289cc1bf2/Inspectors/UX/Pointers/ParabolicTeleportPointerInspector.cs:28)
UnityEditor.UIElements.InspectorElement+<CreateIMGUIInspectorFromEditor>c__AnonStorey1.<>m__0 () (at C:/buildslave/unity/build/Editor/Mono/Inspector/InspectorElement.cs:462)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```
thrown at each frame

## Screenshots

Screenshot (78)

## Your Setup (please complete the following information)

- Unity Version 2019.1.14f1
- XRTK Version e.g. 0.1.8

## Target Platform (please complete the following information)

- WMR immersive
- OpenVR

## Additional context

- HP Windows Mixed Reality headset
- Win 1903

### LINKED ISSUE #90: MixedRealityInputSystem error; Failed to start the Input System! No controllers visible
## Describe the bug

After Mixed Reality Toolkit Configure errors appear. Unity Plays but no controllers are visible

```
Failed to register the MixedRealityInputSystem service: System.Exception - The Input system is missing the required GazeProviderType!
  at XRTK.Services.InputSystem.MixedRealityInputSystem..ctor (XRTK.Definitions.InputSystem.MixedRealityInputSystemProfile profile) [0x000da] in D:\M\Template\XRTKTemplate12b\Library\PackageCache\com.xrtk.core@0.1.20\Services\InputSystem\MixedRealityInputSystem.cs:44 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in <7d97106330684add86d080ecf65bfe69>:0 
UnityEngine.Debug:LogError(Object)
XRTK.Services.MixedRealityToolkit:CreateAndRegisterService(Type, SupportedPlatforms, Object[]) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:870)
XRTK.Services.MixedRealityToolkit:CreateAndRegisterService(Type, Object[]) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:820)
XRTK.Services.MixedRealityToolkit:InitializeServiceLocator() (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:368)
XRTK.Services.MixedRealityToolkit:ResetConfiguration(MixedRealityToolkitConfigurationProfile) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:117)
XRTK.Inspectors.MixedRealityToolkitInspector:OnInspectorGUI() (at Library/PackageCache/com.xrtk.core@0.1.20/Inspectors/MixedRealityToolkitInspector.cs:99)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Failed to start the Input System!
UnityEngine.Debug:LogError(Object)
XRTK.Services.MixedRealityToolkit:InitializeServiceLocator() (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:403)
XRTK.Services.MixedRealityToolkit:ResetConfiguration(MixedRealityToolkitConfigurationProfile) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:117)
XRTK.Inspectors.MixedRealityToolkitInspector:OnInspectorGUI() (at Library/PackageCache/com.xrtk.core@0.1.20/Inspectors/MixedRealityToolkitInspector.cs:99)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Unable to register a service with a null concrete IMixedRealitySpatialObserverDataProvider type.
UnityEngine.Debug:LogError(Object)
XRTK.Services.MixedRealityToolkit:CreateAndRegisterService(Type, SupportedPlatforms, Object[]) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:852)
XRTK.Services.MixedRealityToolkit:InitializeServiceLocator() (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:434)
XRTK.Services.MixedRealityToolkit:ResetConfiguration(MixedRealityToolkitConfigurationProfile) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:117)
XRTK.Inspectors.MixedRealityToolkitInspector:OnInspectorGUI() (at Library/PackageCache/com.xrtk.core@0.1.20/Inspectors/MixedRealityToolkitInspector.cs:99)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Failed to start Lumin Spatial Observer Data Provider!
UnityEngine.Debug:LogError(Object)
XRTK.Services.MixedRealityToolkit:InitializeServiceLocator() (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:441)
XRTK.Services.MixedRealityToolkit:ResetConfiguration(MixedRealityToolkitConfigurationProfile) (at Library/PackageCache/com.xrtk.core@0.1.20/Services/MixedRealityToolkit.cs:117)
XRTK.Inspectors.MixedRealityToolkitInspector:OnInspectorGUI() (at Library/PackageCache/com.xrtk.core@0.1.20/Inspectors/MixedRealityToolkitInspector.cs:99)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

Additional error that appearted once Core has been added
```
SocketException: An attempt was made to access a socket in a way forbidden by its access permissions.

System.Net.Sockets.Socket.Bind (System.Net.EndPoint localEP) (at <0079a30f96a047348857e1cecc6c638a>:0)
SyntaxTree.VisualStudio.Unity.Messaging.UdpSocket.Bind (System.Net.IPAddress address, System.Int32 port) (at <7bfe50ea87ae444bb97944401234ceb7>:0)
SyntaxTree.VisualStudio.Unity.Messaging.Messager..ctor (System.Int32 port) (at <7bfe50ea87ae444bb97944401234ceb7>:0)
SyntaxTree.VisualStudio.Unity.Messaging.Messager.BindTo (System.Int32 port) (at <7bfe50ea87ae444bb97944401234ceb7>:0)
SyntaxTree.VisualStudio.Unity.Bridge.VisualStudioIntegration+<>c__DisplayClass9_0.<.cctor>b__0 () (at <843b7c5ab4d34b2491669c24f4d94522>:0)
SyntaxTree.VisualStudio.Unity.Bridge.Project+<>c__DisplayClass40_0.<RunOnceOnUpdate>b__0 () (at <843b7c5ab4d34b2491669c24f4d94522>:0)
UnityEditor.EditorApplication.Internal_CallUpdateFunctions () (at C:/buildslave/unity/build/Editor/Mono/EditorApplication.cs:200)
```

## To Reproduce
1. Start Hub
2. New Project with Unity 2019.1.14f1; 3D
3. Modify manifest to include 0.1.20
4. Uppon import there is a warning:
`An AssetDatabase Refresh is pending after a successful Package Manager request. You should not call AssetDatabase.Refresh at this time because it can lead to import-related errors.`
It is not clear if one should wait for something
5. Waited for 3 minutes unity manually restarted; warning is gone
6. Through PM added com.xrtk.wmr 0.1.6
7. Toolkit Configure at this point creates Active profile: None
8. Deleted MixedRealityToolkit and MixedRealityPlayspace
9. Imported xrtk.SDK and received errors reported in https://github.com/XRTK/SDK/issues/90
10. Configure successful (after the Unity Crash), but play not possible due to SDK errors
Additional errors appear (listed above)
11. If I remove the SDK through PM the Unity Plays, but these errors remain and controllers are not visible in scene

## Expected behavior
No errors; Motion Controllers registered

## Information(please complete the following information):
- Version 2019.1.14f1
- Release 0.1.20

## Additional context
WMR Headset, OpenVR

### PR DESCRIPTION
**Pull request by StephenHodgson**
## Overview

Continued from https://github.com/XRTK/XRTK-Core/pull/298

## Changes

- Fixes #89 
- Fixes #90

### CODE REVIEW

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/Input/Handlers/ManipulationHandler.cs [hunk 1]
```diff
@@ -346,41 +382,93 @@ protected virtual void Awake()
...
+                    const float nudgeFactor = 0.5f;
+                    float prevExtent;
+                    float newExtent;
+
+                    const float scaleFactor = 0.1f;
+                    Vector3 newScale;
+
+                    switch (keyCode)
+                    {
+                        case KeyCode.W:
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/Input/Handlers/ManipulationHandler.cs]: This was added to help with testing in the editor. It should either be updated to utilize the input actions, or removed.

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/Input/Handlers/ManipulationHandler.cs [hunk 2]
```diff
@@ -394,6 +482,21 @@ protected override void OnDisable()
             }
         }
 
+        protected virtual void OnCollisionEnter(Collision other)
+        {
+            //Debug.LogWarning($"OnCollisionEnter {name} -> {other.gameObject.name}");
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/Input/Handlers/ManipulationHandler.cs]: This collision stuff methods can probably be removed as well

File: XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/BoundingBox/BoundingBox.cs [hunk 3]
```diff
@@ -1204,15 +1251,18 @@ private void UpdateBounds()
                 boundsExtents.z = flattenAxis == FlattenMode.FlattenZ ? 0.0f : boundsExtents.z;
                 currentBoundsExtents = boundsExtents;
 
-                GetCornerPositionsFromBounds(new Bounds(Vector3.zero, boundsExtents * 2.0f), ref boundsCorners);
+                new Bounds(Vector3.zero, boundsExtents * 2f).GetCornerPositionsLocalSpace(ref boundsCorners);
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/BoundingBox/BoundingBox.cs]: this isn't assigned to anything?

Comment (StephenHodgson) [reply XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/BoundingBox/BoundingBox.cs]: Nope it's using the `ref` to update the bounds corners.

**Comment by StephenHodgson** [comment]
Last thing I need to do for this is test that the handles work correctly.

**Comment by StephenHodgson** [comment]
Finally this is ready for review.

**Comment by SimonDarksideJ** [comment]
Code seems fine, but what is the best way to test this in a solution?

**Comment by SimonDarksideJ** [comment]
Getting the following error with Unity
```
Packages\com.xrtk.sdk\Features\Input\Handlers\ManipulationHandler.cs(1106,13): error CS1929: 'Collider' does not contain a definition for 'GetCornerPositionsWorldSpace' and the best extension method overload 'BoundsExtensions.GetCornerPositionsWorldSpace(Bounds, Transform, ref Vector3[])' requires a receiver of type 'Bounds'

```

**Comment by StephenHodgson** [comment]
I'll double check and make sure I committed everything 🤔

**Comment by StephenHodgson** [comment]
Fixed the collider change I made. I guess that'l gonna have to wait for the next release.

To test, 

Create any 3d primitive in a test scene and add a `BoundingBox`.
You'll still need to configure the input actions like select and cancel.
I'd also suggest enabling the `BoundingBox` to activate on enable.

Cheers

**Comment by SimonDarksideJ** [comment]
So is the collider change meant to be in or not. WIll test later when I get back to a pc

**Comment by StephenHodgson** [comment]
It's not that important. Can wait until next release

**Pull request merged**